amr-wind/equation_systems/icns/source_terms/DragForcing.cpp

```diff
@@ -0,0 +1,153 @@
+    bool is_terrain = this->m_sim.repo().field_exists("terrainBlank");
+    if (!is_terrain) {
+        amrex::Abort("Need terrain blanking variable to use this source term");
+    }
+    auto* const m_terrainBlank = &this->m_sim.repo().get_field("terrainBlank");
+    const auto& blank = (*m_terrainBlank)(lev).const_array(mfi);
+    auto* const m_terrainDrag = &this->m_sim.repo().get_field("terrainDrag");
+    const auto& drag = (*m_terrainDrag)(lev).const_array(mfi);
+    auto* const m_terrainz0 = &this->m_sim.repo().get_field("terrainz0");
+    const auto& terrainz0 = (*m_terrainz0)(lev).const_array(mfi);
+    const auto& geom_vec = m_mesh.Geom();
+    const auto& geom = geom_vec[lev];
+    const auto& dx = geom.CellSize();
+    const auto& prob_lo = geom.ProbLoArray();
+    const auto& prob_hi = geom.ProbHiArray();
+    const amrex::Real gpu_drag = m_drag;
+    const amrex::Real gpu_spongeStrength = m_spongeStrength;
+    const amrex::Real gpu_spongeDensity = m_spongeDensity;
+    const amrex::Real gpu_startX = (m_spongeDistanceX > 0)
+                                       ? prob_hi[0] - m_spongeDistanceX
+                                       : prob_lo[0] - m_spongeDistanceX;
+    const amrex::Real gpu_startY = (m_spongeDistanceY > 0)
+                                       ? prob_hi[1] - m_spongeDistanceY
+                                       : prob_lo[1] - m_spongeDistanceY;
+    const amrex::Real gpu_spongeX = (m_spongeDistanceX > 0) ? 1 : 0;
+    const amrex::Real gpu_spongeY = (m_spongeDistanceY > 0) ? 1 : 0;
+    // Copy Data
+    const auto* local_gpu_vel_ht = gpu_vel_ht.data();
+    const auto* local_gpu_vel_vals = gpu_vel_vals.data();
+    const unsigned vsize = gpu_vel_ht.size();
+    amrex::ParallelFor(bx, [=] AMREX_GPU_DEVICE(int i, int j, int k) noexcept {
+        const amrex::Real x1 = prob_lo[0] + (i + 0.5) * dx[0];
+        const amrex::Real x2 = prob_lo[1] + (j + 0.5) * dx[1];
+        const amrex::Real x3 = prob_lo[2] + (k + 0.5) * dx[2];
+        amrex::Real xdamping = 0;
+        amrex::Real ydamping = 0;
+        amrex::Real xi = (gpu_spongeX == 1)
+                             ? (x1 - gpu_startX) / (prob_hi[0] - gpu_startX)
+                             : (gpu_startX - x1) / (gpu_startX - prob_lo[0]);
+        xi = std::max(xi, 0.0);
+        xdamping = gpu_spongeStrength * xi * xi;
+        amrex::Real yi = (gpu_spongeY == 1)
+                             ? (x2 - gpu_startY) / (prob_hi[1] - gpu_startY)
+                             : (gpu_startY - x2) / (gpu_startY - prob_lo[1]);
+        yi = std::max(yi, 0.0);
+        ydamping = gpu_spongeStrength * yi * yi;
+        amrex::Real Cd = gpu_drag / dx[0];
+        amrex::Real gpu_spongeVelX = 0.0;
+        amrex::Real gpu_spongeVelY = 0.0;
+        amrex::Real gpu_spongeVelZ = 0.0;
+        amrex::Real residual = 1000;
+        amrex::Real height_error = 0.0;
+        for (unsigned ii = 0; ii < vsize; ++ii) {
```

Comment:
There might be a better way to find the first closest index to the reference height `x3` using `std::lower_bound` and `std::upper_bound` or some such instead of this loop?

Reply:
Completely agreed,  forgot to update in the PR, but I clarified with @hgopalan offline to not use the `std` functions right after I had this comment here.